src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -1,15 +1,111 @@
+/*
+ * Copyright 2009 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.gradle.java;
 
+import org.gradle.api.Action;
+import org.gradle.api.GradleException;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.file.CopySpec;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.logging.Logging;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
+import org.gradle.api.tasks.compile.JavaCompile;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
 
 public class JigsawPlugin implements Plugin<Project> {
     private static final Logger LOGGER = Logging.getLogger(JigsawPlugin.class);
 
+    private static final String JAVA_PLUGIN = "org.gradle.java";
+
+    private static final String JAVA_LIBRARY_PLUGIN = "org.gradle.java-library";
+
+    private static final String APPLICATION_PLUGIN = "application";
+
     @Override
     public void apply(Project project) {
         LOGGER.debug("Applying JigsawPlugin to " + project.getName());
+
+        project.getExtensions().create("jigsawModule", JigsawModule.class);
+
+        configureJavaTasks(project);
+    }
+
+    private void configureJavaTasks(Project project) {
+        project.afterEvaluate(new Action<Project>() {
+            @Override
+            public void execute(Project project) {
+                if (project.getPlugins().hasPlugin(JAVA_PLUGIN) || project.getPlugins().hasPlugin(JAVA_LIBRARY_PLUGIN)) {
+                    configureCompileJavaTask(project);
+                } else {
+                    throw new GradleException("You must apply the java or java-library plugin as well.");
+                }
+            }
+        });
+    }
+
+    private void configureCompileJavaTask(final Project project) {
+        final JavaCompile compileJava = getCompileJavaTask(project);
+        final SourceSet main = ((SourceSetContainer) project.getProperties().get("sourceSets")).getByName("main");
+        final JigsawModule module = (JigsawModule) project.getExtensions().getByName("jigsawModule");
+        final String moduleDir = String.format("%s%s%s",
```

Comment:
:x: `String.format` is particularly slow. Should just use good old string concatenation here.